libc/utils/docgen/signal.json

```diff
@@ -1,29 +1,56 @@
 {
-  "macros": [
-    "SIG_DFL",
-    "SIG_ERR",
-    "SIG_IGN",
-    "SIGABRT",
-    "SIGFPE",
-    "SIGILL",
-    "SIGINT",
-    "SIGSEGV",
-    "SIGTERM"
-  ],
+  "macros": {
+    "SIG_DFL": {
+      "defined": "7.14.3",
+      "implemented": false
+    },
+    "SIG_ERR": {
+      "defined": "7.14.3",
+      "implemented": false
+    },
+    "SIG_IGN": {
+      "defined": "7.14.3",
+      "implemented": false
+    },
+    "SIGABRT": {
+      "defined": "7.14.3",
+      "implemented": true
+    },
+    "SIGFPE": {
+      "defined": "7.14.3",
+      "implemented": true
+    },
+    "SIGILL": {
+      "defined": "7.14.3",
+      "implemented": true
+    },
+    "SIGINT": {
+      "defined": "7.14.3",
+      "implemented": true
+    },
+    "SIGSEGV": {
+      "defined": "7.14.3",
+      "implemented": true
+    },
+    "SIGTERM": {
+      "defined": "7.14.3",
+      "implemented": true
+    }
+  },
   "functions": {
+    "signal": {
+      "defined": "7.14.1.1"
+    },
+    "raise": {
+      "defined": "7.14.2.1"
+    },
```

Comment:
> I guess a lot of bloviating, but all that to ask what is your reasoning here for moving signal+raise and do we need to change the current sort here in this PR?

I have no strong opinions on the orderings. Based on the previous files, I've been keeping the json files in order of occurrence in the C standard for ease of auditing alongside the standard. This is why I bumped signal and raise up, I was going to let the POSIX functions fall after but totally fine with reverting or alphabetically sorting. `docgen.py` is sorting everything in the `rst` files alphabetically by function/macro name currently. Order in the rst docs seems more, 'how will people use it'?

> Due to 2, these should ultimately be a list, or perhaps another object with distinct fields for C standard section number and hyperlink to POSIX docs (or linux man pages, for other APIs even).

Yeah, I haven't started on listing POSIX hyperlinks yet. I was leaning towards the latter solution with distinct fields and changing `docgen.py` to: change the "Standard" rst column to "C23 Standard", add an optional `POSIX standard#` column (do we need the standard numbers to differentiate C/POSIX standard versions?) that is only added to the generated rst if some row has a POSIX reference in the json.

Reply:
>  I was going to let the POSIX functions fall

POSIX-only

> docgen.py is sorting everything in the rst files alphabetically by function/macro name currently.

Yeah, I noticed.  I suspect the json parser is using ordered dictionaries.

I think my feeling here is: if it's easy to sort via automation, keep it sorted. Otherwise, whatever. At least for the config files.

> Order in the rst docs seems more, 'how will people use it'?

Right. Let's wait to collect feedback from more consumers before making a decision.  We could sort by C standard section for instance.  Then perhaps POSIX sections.  Let's hold on to that thought until later though.

---

> I was leaning towards the latter solution with distinct fields

SGTM

> do we need the standard numbers to differentiate C/POSIX standard versions?

So far, we're only tracking the latest accepted C standard.  You'll notice the sections jump around between C standard revisions because new functions get added in the middle of existing sections.

So I think we should do the same for POSIX and just refer to POSIX.1-2017.

IIUC some functions were deprecated by newer POSIX standards, but many programs won't be rewritten to reflect that, so most libc's need to provide even deprecated POSIX functionality to be somewhat portable-to.  IDK if POSIX has ever removed anything; in that case we'd have to point back to prior POSIX releases.

Let's ask the maintainer of bionic (Android's libc), @enh-google , if he knows answers to the above?

To me, currently, docgen is more about:
1. reflecting to users how "done" we are via some form of visualization (even if fragmented across numerous html pages)
2. visualizing what could be implemented next, either by the core team or new contributors.
3. helping implementers quickly know which section of which doc they should read to understand the semantics required by "the spec" for whichever specs are relevant.

Maybe someday docgen grows into more (such as when these things were first standardized), but I'd rather folks contribute to the upstream man pages project as duplicating what they have is frankly a waste of time and is user hostile.

Reply:
> Yeah, I noticed. I suspect the json parser is using ordered dictionaries.

I thought the current sorting in the generated rst docs was on purpose, there is an [explicit sort on function names](https://github.com/llvm/llvm-project/blob/22cd7f28164543895eef245b30348427a379a7be/libc/utils/docgen/docgen.py#L47).

Another ordering to take into account, that does bug me a bit, is the ordering of the listed headers in the left side bar on libc.llvm.org: "Search Tables" and "C23 Support" in the middle of the standard headers. Thoughts on reordering these to push the standard header status to the bottom on the side bar and alphabetically sorted?

Reply:
> IIUC some functions were deprecated by newer POSIX standards, but many programs won't be rewritten to reflect that, so most libc's need to provide even deprecated POSIX functionality to be somewhat portable-to. IDK if POSIX has ever removed anything; in that case we'd have to point back to prior POSIX releases.
> 
> Let's ask the maintainer of bionic (Android's libc), @enh-google , if he knows answers to the above?

yes, POSIX has removed things. no, most C libraries do not remove those things.

i regret all of the functions i removed. it didn't gain me anything (because we'd already done the work), and it hurt developers (if only slightly).

that said, i do _not_ regret the ones i never added https://android.googlesource.com/platform/bionic/+/main/docs/status.md#posix --- POSIX is full of useless crap while also missing lots of essential stuff. (theoretically that makes sense viewed from their "codifying existing practice" stance, but at the same time they remove tar(1) in favor of pax(1), or add gzip to compress(1) rather than add gzip(1), so their grasp of "existing practice" is tenuous at best.)

Reply:
> I thought the current sorting in the generated rst docs was on purpose, there is an [explicit sort on function names](https://github.com/llvm/llvm-project/blob/22cd7f28164543895eef245b30348427a379a7be/libc/utils/docgen/docgen.py#L47).

Oh, that's right. I wrote that.  Either way, I don't have my heart set on that.  What's more important to me is that we maintain some form of determinism in output in case folks decide to re-arrange the order of things in json.  Though maybe we just keep the json sorted (painful) and rely on that ordering.  Personally, I think it'll be easier to enforce the ordering if the tool just sorts every time, then we can add new stuff to the json in any order.

> Another ordering to take into account, that does bug me a bit, is the ordering of the listed headers in the left side bar on libc.llvm.org

heh, I was just talking with @dpxf about that.  yeah it's ugly and will grow to be unwieldy very quickly.  I'd like to have one tab, that links to a newly created page, that lists each header we provide, which is a link to the corresponding page.